### ISSUE TITLE
Upload chaincode to nodes

### ISSUE THREAD

**Issue by yeasy** [hyperledger-archives/fabric#1381]
Hi,
    I notice we can only deploy the existed code within the file system in the nodes.
    Is there any way we can upload new chaincode file into the node?
    Should we have such kind of API in the fabric?
    Thanks!

**Comment by mastersingh24** [comment]
@yeasy - trying to understand the exact requirement here.  Currently, you can:
- have your code locally under the GOPATH and use the peer CLI to deploy to a node in the network
- use the REST API to deploy code which is either in a public github repo or exists locally on the peer whose API endpoint you are targeting

I assume you are asking for an additional method which allows you to package up code and send it to a peer?

**Comment by ghaskins** [comment]
In addition to the methods @mastersingh24 mentioned, I am also planning on proposing an "inline" option where the chaincode payload may be submitted in the API parameters.  I imagine this method will make more sense to chaincodeType=CAR packages as emitted by [chaintool](https://github.com/ghaskins/chaintool), but there may be other future chaincodeTypes that could benefit as well.

**Comment by smithbk** [comment]
The client SDK will allow passing bytes to deploy.  The gRPC call already allows it but not exposed via REST.  See codePackage below.  Will that meet the need here?

message ChaincodeDeploymentSpec {

```
enum ExecutionEnvironment {
    DOCKER = 0;
    SYSTEM = 1;
}

ChaincodeSpec chaincodeSpec = 1;
// Controls when the chaincode becomes executable.
google.protobuf.Timestamp effectiveDate = 2;
bytes codePackage = 3;
ExecutionEnvironment execEnv=  4;
```

}

**Comment by yeasy** [comment]
@mastersingh24 @ghaskins @smithbk 
Thanks all for the comment!
I guess we at least should support the following two ways to deploy a chaincode addtionally:
1. given a public URL, and peer node will download and deploy it. (I do not find an example for this however @mastersingh24 )
2. Send codechain in payload with plain or base64 code.

Moreover, we can upload local codechain file to the peer.

**Comment by ghaskins** [comment]
@yeasy note that item 1 is already supported today (use chaincodeid.path with a URL).  Item 2 is not supported (yet)

**Comment by muralisrini** [comment]
@smithbk  ChaincodeDeploymentSpec.CodePackage contains bytes suitable for deployment (e.g., targz for Docker) computed from the source code pointed to by ChaincodeID.Path.  

IIUC... SDK will be sending protos.Transaction to peer ?   In that case we could just move that computation to the SDK. We wouldn't overload the semantics of CodePackage field.

**Comment by christo4ferris** [comment]
@smithbk unless I am mistaken, we an do both 1 and 2 from @yeasy's comment above. If so, please advise and we can close this. We have #873 to track access to a private repo, we don't need this to reinforce that especially since it references this one.

**Comment by wdawson4** [comment]
@yeasy see #873 for a potential solution to deploying from a private repository

**Comment by angrbrd** [comment]
@yeasy @christo4ferris 

Currently, we are able to deploy from a public URL through the REST API only. The SDK does not have a feature of deploying from a public URL. The SDK only deploys from a local chaincode directory on the client machine by packaging the chaincode .tar.gz for Docker and sending that in the transaction payload.

Are we looking to add a functionality to deploy from a public URL through the SDK to address this feature? Thanks.

**Comment by smithbk** [comment]
Yes, I'm working on adding deploying zip and targz files by URL or local file.  They may contain the Dockerfile already and if not it will be added by the node SDK at deploy time.

**Comment by angrbrd** [comment]
This issue is already being tracked in Jira here: https://jira.hyperledger.org/browse/FAB-80

@christo4ferris please close this issue. Thanks!
